src/main/java/com/linecorp/armeria/server/metrics/ServiceMetricDecorator.java

```diff
@@ -0,0 +1,152 @@
+
+    private final ServiceMetricConsumer metricConsumer;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ServiceCodec} with
+     * the specified {@link ServiceMetricConsumer}.
+     */
+    public ServiceMetricDecorator(ServiceCodec codec, ServiceMetricConsumer consumer) {
+        super(codec);
+        metricConsumer = consumer;
+    }
+
+    @Override
+    public DecodeResult decodeRequest(Channel ch, SessionProtocol sessionProtocol, String hostname, String path,
+                                      String mappedPath, ByteBuf in, Object originalRequest,
+                                      Promise<Object> promise) throws Exception {
+        final long startTime = System.nanoTime();
+        final int requestSize = in.readableBytes();
+
+        DecodeResult decodeResult = delegate().decodeRequest(
+                ch, sessionProtocol, hostname, path, mappedPath, in, originalRequest, promise);
+
+        LongSupplier lazyElapsedTime = () -> System.nanoTime() - startTime;
+
+        switch (decodeResult.type()) {
+        case SUCCESS: {
+            ServiceInvocationContext context = decodeResult.invocationContext();
+            context.attr(METRICS).set(new Metrics(sessionProtocol, requestSize, startTime));
+
+            promise.addListener(future -> {
+                if (!future.isSuccess()) {
+                    // encodeFailureResponse will process this case.
+                    return;
+                }
+                Object result = future.getNow();
+
+                if (result instanceof FullHttpResponse) {
+                    FullHttpResponse httpResponse = (FullHttpResponse) result;
+                    metricConsumer.invocationComplete(
+                            sessionProtocol, httpResponse.status().code(),
+                            lazyElapsedTime.getAsLong(), requestSize, httpResponse.content().readableBytes());
+                }
+                // encodeResponse will process this case.
+            });
+            break;
+        }
+        case FAILURE: {
+            final Object errorResponse = decodeResult.errorResponse();
+            if (errorResponse instanceof FullHttpResponse) {
+                FullHttpResponse httpResponse = (FullHttpResponse) errorResponse;
+                metricConsumer.invocationComplete(
+                        sessionProtocol, httpResponse.status().code(), lazyElapsedTime.getAsLong(),
+                        requestSize, httpResponse.content().readableBytes());
+            } else {
+                metricConsumer.invocationComplete(
+                        sessionProtocol, HttpResponseStatus.BAD_REQUEST.code(),
+                        lazyElapsedTime.getAsLong(), requestSize, 0);
+            }
+            break;
+        }
+        case NOT_FOUND:
+            metricConsumer.invocationComplete(
+                    sessionProtocol, HttpResponseStatus.NOT_FOUND.code(),
```

Comment:
When `DecodeResultType` is `NOT_FOUND`, the session layer `HttpServerHandler` will keep trying other services until there's a service that does not return `NOT_FOUND`. i.e. You should ignore this `DecodeResultType`.